src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy

```diff
@@ -0,0 +1,353 @@
+import org.junit.Rule
+import org.junit.rules.TemporaryFolder
+import spock.lang.Specification
+
+class ProtobufModelPluginTest extends Specification {
+  @Rule
+  final TemporaryFolder testProjectDir = new TemporaryFolder()
+  File buildFile
+
+  def setup() {
+    buildFile = testProjectDir.newFile('build.gradle')
+
+    def pluginClasspathResource = getClass().classLoader.findResource("plugin-classpath.txt")
+    if (pluginClasspathResource == null) {
+      throw new IllegalStateException("Did not find plugin classpath resource, run `testClasses` build task.")
+    }
+
+    def pluginClasspath = pluginClasspathResource.readLines()
+        .collect { it.replace('\\', '\\\\') } // escape backslashes in Windows paths
+        .collect { "'$it'" }
+        .join(", ")
+
+    // Add the logic under test to the test build
+    buildFile << """
+          buildscript {
+            dependencies {
+              classpath files($pluginClasspath)
+            }
+          }
+        """
+
+    // Generate some useful path
+    def userDir = System.getProperty("user.dir").replace("\\", "\\\\")
+    def vendorDir = "$userDir/vendor"
+    def vendorBinaryDir = "$vendorDir/build/binaries"
+
+    // Check to see if the vendor binaries were generated prior to running the test
+    assert new File("$vendorBinaryDir/protocExecutable").exists()
+    assert new File("$vendorBinaryDir/protobufStaticLibrary").exists()
+    assert new File("$vendorBinaryDir/protobufLiteStaticLibrary").exists()
+
+    buildFile << """
+          apply plugin: 'cpp'
+
+          // Repositories doesn't seems to be working for this.
+          binaries.all {
+            if (toolChain instanceof VisualCpp) {
+              cppCompiler.args "/I$vendorDir/protobuf/src"
+              linker.args "/LIBPATH:$vendorBinaryDir/protobufStaticLibrary", "protobuf.lib"
+              linker.args "/LIBPATH:$vendorBinaryDir/protobufLiteStaticLibrary", "protobufLite.lib"
+            } else {
+              cppCompiler.args "-I$vendorDir/protobuf/src"
+              linker.args "-L$vendorBinaryDir/protobufStaticLibrary", "-lprotobuf"
+              linker.args "-L$vendorBinaryDir/protobufLiteStaticLibrary", "-lprotobufLite"
+            }
+          }
+
+          import com.google.protobuf.gradle.*
+          apply plugin: 'com.google.protobuf'
+
+          model {
+            protobuf {
+              protoc {
+                path = "$vendorBinaryDir/protocExecutable/protoc"
+              }
+            }
+          }
+        """
+
+    testProjectDir.newFolder('src', 'main', 'cpp')
+    testProjectDir.newFolder("src", "main", "proto")
+  }
+
+  def "a default protobuf source set exists for configuration"() {
+    given: "a native executable with proto source set configuration"
+    buildFile << """
+          model {
+            components {
+              main(NativeExecutableSpec) {
+                sources {
+                  proto {
+                    source.include "*.proto"
```

Comment:
The include `*.proto` is actually not needed here. I think I wanted to be implicit. The source set follow the same pattern as every other source set in Gradle meaning by default it will include any file under the language directory. The `*.proto` is a configuration taste for the user. I will remove where it's not needed so it doesn't confuse anyone.